### ISSUE TITLE
[no-explicit-any] False positive for spread parameter types (`extends any[]`)

### ISSUE THREAD

**Issue by LinqLover** [typescript-eslint/typescript-eslint#3464]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```yml
rules:
  "@typescript-eslint/no-explicit-any": error
```

```TS
function foo<T extends any[]>(fn: (...args: T) => void, ...args: T): void
```

[Playground](https://yeonjuan.github.io/typescript-eslint-demo/#e2NvZGU6J2Z1bmN0aW9uIGZvbzxUIGV4dGVuZHMgYW55W10+KGZuOiAoLi4uYXJnczogVCkgPT4gdm9pZCwgLi4uYXJnczogVCk6IHZvaWQnLHJ1bGVzOnsnQHR5cGVzY3JpcHQtZXNsaW50L25vLWV4cGxpY2l0LWFueSc6J2Vycm9yJ30scGFyc2VyT3B0aW9uczp7ZWNtYUZlYXR1cmVzOntqc3g6ZmFsc2UsZ2xvYmFsUmV0dXJuOmZhbHNlfSxlY21hVmVyc2lvbjoyMDE4LHByb2plY3Q6WycuL3RzY29uZmlnLmpzb24nXSxzb3VyY2VUeXBlOidzY3JpcHQnfX0=)

**Expected Result**

There should be no linter warning/error.

**Actual Result**

The `no-explicit-any` rule is violated.

**Additional Info**

Despite the general use of this rule, from what I found out there is no equivalent alternative to the `extends any[]` syntax which is used for variadic parameter types in TypeScript.

Actually, this syntax does not add any arbitrariness to the input/output types but it is still type-safe. You can convince yourself of this fact by adding the following statement to the declaration:

```ts
foo((x: number, y: string) => 42, 'foo', 42) // TS2345
```

For this reason, I think that `extends any[]` should not trigger this rule.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^4.22.1` |
| `@typescript-eslint/parser`        | `^4.22.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.12.1` |

**Comment by LinqLover** [comment]
Just found <https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md#ignorerestargs>. Nevertheless, given my argument from above, why do we need this option at all?

**Comment by bradzacher** [comment]
Some people don't want any `any`s in their codebase at all.
In a lot of codebase when people reach for these sorts of signatures they can go with something more specific - so it's a good option to have.

The option is going to remain. If you want to use the signature, then use a disable comment or turn the option on.

**Comment by LinqLover** [comment]
Thank you for the reply! Just for sake of completeness ...

> In a lot of codebase when people reach for these sorts of signatures they can go with something more specific - so it's a good option to have.

Could you maybe provide a short example of how to forego the `any` statement in my example from above? :-)

**Comment by bradzacher** [comment]
A lot of times people go "I want a function which can accept any function with any arguments", but they don't actually mean that. They're thinking abstractly about an infinitely reusable utility function, but they only actually need to solve for one or two cases.

There certainly are cases you need the generic signature using any, but usually you don't.
